Develop2

AnatoliyVS79 opened the pull request:
# Diplom_2
Нужно протестировать ручки API для Stellar Burgers.

-Создание пользователя.-
-Логин пользователя.-
-Изменение данных пользователя.-
-Создание заказа.-
-Получение заказов конкретного пользователя.-

Стэк для тестирования:
-- Java 11
-- junit 4.13.2
-- RestAssured 5.0.1
-- Allure 2.17.3.

- Команда для запуска тестов: mvn clean test.
- Для запустится веб-сервер Allure ввести команду: mvn allure:serve

AnatoliyVS79 committed:
added file pom.xml

AnatoliyVS79 committed:
added BaseData.java Token.java User.java

AnatoliyVS79 committed:
added UserLogin.java  UserRequest.java

AnatoliyVS79 committed:
added Data.java IngredientList.java

AnatoliyVS79 committed:
added Order.java

AnatoliyVS79 committed:
added NewUserTest.java

AnatoliyVS79 committed:
added LoginUserTest.java

AnatoliyVS79 committed:
added ChangingUserDataTest.java

AnatoliyVS79 committed:
added CreateOrderTest

AnatoliyVS79 committed:
added GetUserOrderTest

AnatoliyVS79 committed:
change pom.xml , README.md

AnatoliyVS79 committed:
add allure report

AnatoliyVS79 committed:
change .gitignore

pomegranate282922 commented on src/main/java/ru/praktikum/stellarburgers/client/IngredientList.java:
```diff
@@ -0,0 +1,50 @@
+        return ingredientList;
+    }
+
+    String pp = ",";
```
Сначала перечисляем все поля в классе, потом все методы. Нужно исправить

pomegranate282922 commented on src/main/java/ru/praktikum/stellarburgers/client/IngredientList.java:
```diff
@@ -0,0 +1,50 @@
+
+import static io.restassured.RestAssured.given;
+
+public class IngredientList extends BaseData {
```
Нужно разнести : отдельно оставить pojo. И отдельно сделать класс с вызовами ручек. Нужно исправить.

pomegranate282922 replied:
Не исправлено. Вызов ручки нужно отделить от pojo - разнести их по разным классам. Нужно исправить

AnatoliyVS79 replied:
Переделал.

pomegranate282922 replied:
Не увидела(

pomegranate282922 commented on src/main/java/ru/praktikum/stellarburgers/client/Order.java:
```diff
@@ -0,0 +1,53 @@
+import static io.restassured.RestAssured.given;
+
+public class Order extends BaseData {
+    public static final String ORDER_PATH = BASE_URL + "orders/";
```
Здесь и везде base url не нужен, так как уже установлен в req spec. Можно улучшить

pomegranate282922 commented on src/main/java/ru/praktikum/stellarburgers/client/Order.java:
```diff
@@ -0,0 +1,53 @@
+                .get(ORDER_PATH);
+    }
+
+    @Step("Вывод заказов неавторизованного пользователя")
```
Отлично, что есть аллюр степы

pomegranate282922 commented on src/main/java/ru/praktikum/stellarburgers/model/BaseData.java:
```diff
@@ -0,0 +1,16 @@
+public class BaseData {
+    public static final String BASE_URL = "https://stellarburgers.nomoreparties.site/api/";
+
+    protected static RequestSpecification getBaseSpec(){
```
Отлично, что используется req spec. Это здорово уменьшает дублирование кода

pomegranate282922 commented on src/main/java/ru/praktikum/stellarburgers/model/User.java:
```diff
@@ -0,0 +1,34 @@
+import org.apache.commons.lang3.RandomStringUtils;
+
+public class User {
+    public String login;
```
Здесь нарушение принципа инкапсуляции. Поля в pojo делаем приватными, доступ через геттер.

pomegranate282922 commented on src/main/java/ru/praktikum/stellarburgers/model/User.java:
```diff
@@ -0,0 +1,34 @@
+        this.name = name;
+    }
+
+    public static User createUser(){
```
Отлично, что данные рандомизированы

pomegranate282922 commented on src/main/java/ru/praktikum/stellarburgers/model/UserLogin.java:
```diff
@@ -0,0 +1,22 @@
+package ru.praktikum.stellarburgers.model;
+
+public class UserLogin {
+    public String login;
```
Нарушение принципа инкапсуляции нужно исправить везде

pomegranate282922 commented on src/main/java/ru/praktikum/stellarburgers/model/UserRequest.java:
```diff
@@ -0,0 +1,52 @@
+                .post(USER_PATH + "register/");
+    }
+
+    @Step("Регистрация пользователя {userLogin}")
```
Логин это не регистрация. Логин это вход в систему

pomegranate282922 commented on src/test/java/ru/yandex/stellarburgers/ChangingUserDataTest.java:
```diff
@@ -0,0 +1,102 @@
+    @DisplayName("Изменение email авторизированного пользователя")
+    @Description("Тест проверяет возможность изменения email авторизированного пользователя")
+    public void userChangeEmail(){
+        String changeData = "{\"" + "email" + "\":" + "\"" + RandomStringUtils.randomAlphabetic(3) +
```
Здесь и везде прошу поправить. Передавать тело запроса в виде строки неудобно и нечитаемо. Нужно использовать либо встроенную в RestAssured сериализацию, либо передавать тело как мапу.
Нужно исправить везде

pomegranate282922 replied:
Не исправлено. Не передаём тело в виде строки - передаём через объект или мапу. Нужно исправить

AnatoliyVS79 replied:
Переделал всё.

pomegranate282922 replied:
Не вижу изменений. Json как передавался в виде строки, так и передаётся. Нужно исправить везде

pomegranate282922 commented on src/test/java/ru/yandex/stellarburgers/NewUserTest.java:
```diff
@@ -0,0 +1,88 @@
+    @DisplayName("Создание пользователя без указания поля password")
+    @Description("Тест проверяет появление ошибки в случае отсутствия обязательного поля password")
+    public void userNotBeCreateWithoutRequiredFieldPassword(){
+        User user2 = new User(user.login, "", user.name);
```
Не используем магические цифры в именах. Нужно исправить.
Более того, здесь лучше использовать user, чтобы в after он удалился, если тест поймает баг и юзер создастся 
Нужно исправить

pomegranate282922 commented on src/test/java/ru/yandex/stellarburgers/NewUserTest.java:
```diff
@@ -0,0 +1,88 @@
+        userRequest.create(user2.toString()).then().assertThat()
+                .statusCode(403)
+                .body("message", Matchers.equalTo("Email, password and name are required fields"));
+    }
```
Аналогично

pomegranate282922 reviewed (commented):
Работа хорошая! Тесты все нужные реализованы, отлично, что есть рандомизация данных.
Немного набралось замечаний - прошу их поправить.

AnatoliyVS79 committed:
added UserRegResponse

AnatoliyVS79 committed:
correction of comments

AnatoliyVS79 committed:
add allure report

AnatoliyVS79 committed:
add allure report

pomegranate282922 commented on src/main/java/ru/praktikum/stellarburgers/client/BaseData.java:
```diff
@@ -1,13 +1,13 @@
 
 public class BaseData {
-    public static final String BASE_URL = "https://stellarburgers.nomoreparties.site/api/";
+    public final String BASE_URL = "https://stellarburgers.nomoreparties.site/";
```
Зачем убран статик? Нужно исправить

AnatoliyVS79 replied:
Переделал

pomegranate282922 commented on src/main/java/ru/praktikum/stellarburgers/client/UserRequest.java:
```diff
@@ -1,31 +1,36 @@
-public class UserRequest extends BaseData{
-    public static final String USER_PATH = BASE_URL + "auth/";
+public class UserRequest extends BaseData {
+    public  final String USER_PATH = "api/auth/";
```
Зачем отовсюду убран статик? Константы являются статиками. Нужно исправить.

AnatoliyVS79 replied:
Переделал

pomegranate282922 commented on src/test/java/ru/yandex/stellarburgers/CreateOrderTest.java:
```diff
@@ -0,0 +1,87 @@
+    @Description("Тест проверяет возврат ошибки в случае передачи запроса о создании заказа без ингредиентов " +
+            "(у неавторизованного пользователя запрос отрабатывает так же)")
+    public void createOrderWithoutIngredientsTest(){
+        order.json = "{" + "\"ingredients\":[]" + "}";
```
Везде нужно убрать передачу тела в виде строки. Это слишком нечитаемо и есть более удобные методы для этого.

pomegranate282922 reviewed (commented):
Многое поправил, но не все - остались важные замечания, которые необходимо поправить.

AnatoliyVS79 committed:
added GetCustomUsersOrders

AnatoliyVS79 committed:
added Request and CreateOrderRequest

AnatoliyVS79 committed:
added LoginUserRequest and UpdateUserRequest

AnatoliyVS79 committed:
class change BaseData, CreateOrderTest, Order, Tokens, User, and UserRegisterResponse

AnatoliyVS79 committed:
added Utils

AnatoliyVS79 committed:
added RegisterUserRequest

AnatoliyVS79 committed:
After code review fix

AnatoliyVS79 committed:
gitignore change

pomegranate282922 commented on src/main/java/ru/praktikum/stellarburgers/client/IngredientList.java:
```diff
@@ -0,0 +1,54 @@
+
+
+    @Step("Вывод списка всех ингредиентов в виде объекта {IngredientList}")
+    public IngredientList getAllIngredients(){
```
Вот это нужно вынести в класс Order, потому что там лежат все вызовы ручек. Нужно исправить

AnatoliyVS79 replied:
Класса IngredientList нет, он удалён

pomegranate282922 commented on src/main/java/ru/praktikum/stellarburgers/model/CreateOrderRequest.java:
```diff
@@ -0,0 +1,9 @@
+package ru.praktikum.stellarburgers.model;
+
+public class CreateOrderRequest extends Request {
+    public final String[] ingredients;
```
Нарушение принципа инкапсуляции. Поля приватные, доступ через геттеры и сеттеры. Нужно исправить

pomegranate282922 commented on src/main/java/ru/praktikum/stellarburgers/model/LoginUserRequest.java:
```diff
@@ -0,0 +1,11 @@
+package ru.praktikum.stellarburgers.model;
+
+public class LoginUserRequest extends Request {
+    public final String email;
```
Аналогично нужно исправить

pomegranate282922 commented on src/main/java/ru/praktikum/stellarburgers/model/RegisterUserRequest.java:
```diff
@@ -0,0 +1,22 @@
+import org.apache.commons.lang3.RandomStringUtils;
+
+public class RegisterUserRequest extends Request {
+    public final String email;
```
Нужно исправить везде в pojo

pomegranate282922 reviewed (commented):
Почему-то некоторые замечания, хотя и отмечены у тебя как выполненные, не устранены( может, забыл в коммит включить? На всякий случай дала ещё пояснения. Нужно исправить все, где есть «нужно исправить».

AnatoliyVS79 committed:
encapsulation principle fixed

AnatoliyVS79 committed:
Deleted classes IngredientList,  BaseData, UserLogin, UserRequest, GetUserOrderTest,
Class ChangingUserDataTest changed to UpdateUserTest
Class NewUserTest changed to UserRegisterTest
After code review fix

pomegranate282922 commented on src/main/java/ru/praktikum/stellarburgers/client/UserClient.java:
```diff
@@ -0,0 +1,98 @@
+                .and()
+                .body(registerUserRequest.toJsonString())
+                .when()
+                .post("auth/register")
```
Урлы нужно выносить как константы. 
так как не заметила этого в прошлых ревью - оставляю как улучшение.

pomegranate282922 reviewed (approved):
Зачёт!
